src/read_data.h

```diff
@@ -0,0 +1,230 @@
+      keyword = line;
+      if (keyword.compare("Atoms ") || keyword.compare("Velocities ")) 
+        return keyword;
+      else
+        printf("ERROR: Unknown identifier in data file: %s\n", keyword.data());
+    }
+    else
+      continue;
+    
+  }
+}
+
+void read_lammps_header(ifstream &file, System* s)
+{
+  string line;
+  // skip 1st line of file
+  getline(file, line);
+  while(1) {
+    
+    getline(file, line);
+    //skip blank lines
+    int pos = line.find_first_not_of(" \r\t\n");
+    if (pos == string::npos)
+      continue;
+    
+    //ignore anything after # by getting substring till # (takes care of comment lines too)
+    line = line.substr(0, line.find('#'));
+    
+    int natoms, ntypes;
+    double xlo, xhi, ylo, yhi, zlo, zhi;
+    const char* temp = line.data(); //convert to C-string for sscanf utility
+    // search line for header keyword and set corresponding variable
+    if(line.find("atoms") != string::npos) {
+      sscanf(temp, "%i", &natoms);
+      s->N = natoms;
+    }
+    else if(line.find("atom types") != string::npos) {
+      sscanf(temp, "%i", &ntypes);
+      s->ntypes = ntypes;
+    }
+    else if(line.find("xlo xhi") != string::npos) {
+      sscanf(temp, "%lg %lg", &xlo, &xhi);
+      s->domain_x = xhi - xlo;
+    }
+    else if(line.find("ylo yhi") != string::npos) {
+      sscanf(temp, "%lg %lg", &ylo, &yhi);
+      s->domain_y = yhi - ylo;
+    }
+    else if(line.find("zlo zhi") != string::npos) {
+      sscanf(temp, "%lg %lg", &zlo, &zhi);
+      s->domain_z = zhi - zlo;
+      break;
+      //TODO: add support for triclinic boxes
+    }
+  }
+}
+    
+
+void read_lammps_atoms(ifstream &file, System* s)
+{
+  string line;
+  auto x = Cabana::slice<Positions>(s->xvf);
+  auto v = Cabana::slice<Velocities>(s->xvf);
+  auto id = Cabana::slice<IDs>(s->xvf);
+  auto type = Cabana::slice<Types>(s->xvf);
+  auto q = Cabana::slice<Charges>(s->xvf);
+  
+  //skip any empty lines before reading in data
+  while (1) {
+    getline(file, line);
+    if (!line.empty())
+      break;
+  }
+  
+  
+  T_INT id_tmp, type_tmp;
+  T_FLOAT x_tmp, y_tmp, z_tmp, q_tmp; 
+  for (int n=0; n < s->N; n++) {
+    const char* temp = line.data();
+    if (s->atom_style == "atomic") {
+      sscanf(temp, "%i %i %lg %lg %lg", &id_tmp, &type_tmp, &x_tmp, &y_tmp, &z_tmp);
+      id(n) = id_tmp; type(n) = type_tmp; x(n,0) = x_tmp; x(n,1) = y_tmp; x(n,2) = z_tmp;
+      q(n) = 0;
+    }
+    if (s->atom_style == "charge") {
+      sscanf(temp, "%i %i %lg %lg %lg %lg", &id_tmp, &type_tmp, &q_tmp, &x_tmp, &y_tmp, &z_tmp);
+      id(n) = id_tmp; type(n) = type_tmp; q(n) = q_tmp; x(n,0) = x_tmp; x(n,1) = y_tmp; x(n,2) = z_tmp;
+    //getline pushed to the end of loop because line already stores the 1st non-blank line
+    //after exiting while loop
+    getline(file, line); 
+    }
+  }
+}
+
+
+void read_lammps_velocities(ifstream &file, System* s)
+{
+  string line;
+  auto x = Cabana::slice<Positions>(s->xvf);
```

Comment:
Should only need v here (try compiling with -Wall -Wextra -pedantic -Werror and see what you get)